Added support for `consolidate` subpackage functions to accept both in-memory or stored datasets

praneethratna opened the pull request:
Addresses #1129, #1237 and now `consolidate` subpackage functions can accept both in-memory or stored datasets.

CC @leewujung

leewujung commented on echopype/echodata/echodata.py:
```diff
@@ -93,6 +92,8 @@ def cleanup_swap_files(self):
                         v.store for k, v in dask_graph.items() if "original-from-zarr" in k
                     ]
                     fs = zarr_stores[0].fs
+                    from ..utils.io import delete_zarr_store
+
```
Just curious why you decided to move these imports to within the function? Is it because they are only used once?

praneethratna replied:
@leewujung they were causing cyclical import and also the reason you've mentioned since anyway they're being used only once, I don't think we need to define them globally in the file.

leewujung commented on echopype/utils/io.py:
```diff
@@ -414,24 +417,46 @@ def validate_source_ds_da(
+    return source_ds_da_ed, file_type
+
+
+def open_ds_da_ed_path(
+    source_ds_da: Optional[Union[xr.Dataset, xr.DataArray, str, Path]],
+    source_ed: Optional[Union[EchoData, str, Path]],
+    storage_options: Optional[dict],
+) -> Tuple[Optional[Union[xr.Dataset, xr.DataArray]], Optional[EchoData]]:
```
Somehow I feel forcing `source_ed` to always be in the input argument is a little awkward, when you only need to deal with `source_ds_da`. Do you think it will work if the function is changed to take in `source`, `source_type`, and `storage_options`, and depending on `source_type` different open function is used.

praneethratna replied:
> Do you think it will work if the function is changed to take in `source`, `source_type`, and `storage_options`, and depending on `source_type` different open function is used.

@leewujung Yes, I have updated the method which will use different open function depending on `source_type`.

leewujung commented on echopype/utils/io.py:
```diff
@@ -380,29 +382,30 @@ def env_indep_joinpath(*args: Tuple[str, ...]) -> str:
-    source_ds_da: Union[xr.Dataset, xr.DataArray, str, Path], storage_options: Optional[dict]
-) -> Tuple[Union[xr.Dataset, str, xr.DataArray], Optional[str]]:
+def validate_source_ds_da_ed(
+    source_ds_da_ed: Union[xr.Dataset, xr.DataArray, EchoData, str, Path],
```
We can probably change `source_ds_da_ed` to just be `source`? Since the type is pretty clear here.

leewujung reviewed (commented):
Hey @praneethratna : Thanks for the PR! I only have pretty small comments and a question here.

While going through the changes, I noticed that the use of `source_Sv_path` does not make much sense () and that results in the mess in `add_splitbeam_angle` that `source_Sv` is separately validated outside of `open_ds_da_ed_path`. We should probably deal with it in a separate PR, to keep this one tidy.

praneethratna commented:
Hey @leewujung I have addressed the above comments and also included the changes for #1237 in this PR since they are based on these changes. Thanks!

leewujung commented:
Hey @praneethratna : The last changes you made removed the part to save the added `theta` and `phi` variables to an existing dataset. Are you thinking that these variables would just be live in `source_Sv`? Maybe we just remove the `add_angle_to_ds` function and put the content (with necessary modifications) in `add_splitbeam_angle`?

leewujung reviewed (changes requested):
Hey @praneethratna : The last changes you made removed the part to save the added `theta` and `phi` variables to an existing dataset. Are you thinking that these variables would just be live in `source_Sv`?

Maybe we just remove the `add_angle_to_ds` function and put the content (with necessary modifications) in `add_splitbeam_angle`?

praneethratna commented:
> Hey @praneethratna : The last changes you made removed the part to save the added `theta` and `phi` variables to an existing dataset. Are you thinking that these variables would just be live in `source_Sv`?
> 
> Maybe we just remove the `add_angle_to_ds` function and put the content (with necessary modifications) in `add_splitbeam_angle`?

Hey @leewujung I think they should live, but anyway I have moved the content to `add_splitbeam_angle` itself and also removed the `return_dataset` parameter!

leewujung commented:
> Hey @leewujung I think they should live, but anyway I have moved the content to `add_splitbeam_angle` itself and also removed the `return_dataset` parameter!

Ah I see what the confusion may be. What I meant is that the functionality of saving the new variables `phi` and `theta` was removed in your changes, but we actually need the functionality (to update the on-disk dataset with the added variables  `phi` and `theta`, while still being able to access them as `angle_alongship` and `angle_athwartship`).

How about we do the following:
- remove `return_dataset`
- add a new argument `to_disk` (default to `True`), where `to_disk=True` means that the angle will be saved to disk via `mode="a"` and the `source_Sv` dataset lazy-loaded back and returned, and `to_disk=False` means that the angle will just be added to `source_Sv` and returned but not saved to disk.

leewujung reviewed (changes requested):
> Hey @leewujung I think they should live, but anyway I have moved the content to `add_splitbeam_angle` itself and also removed the `return_dataset` parameter!

Ah I see what the confusion may be. What I meant is that the functionality of saving the new variables `phi` and `theta` was removed in your changes, but we actually need the functionality (to update the on-disk dataset with the added variables  `phi` and `theta`, while still being able to access them as `angle_alongship` and `angle_athwartship`).

How about we do the following:
- remove `return_dataset`
- add a new argument `to_disk` (default to `True`), where `to_disk=True` means that the angle will be saved to disk via `mode="a"` and the `source_Sv` dataset lazy-loaded back and returned, and `to_disk=False` means that the angle will just be added to `source_Sv` and returned but not saved to disk.

leewujung commented on echopype/consolidate/api.py:
```diff
@@ -240,19 +255,20 @@ def add_splitbeam_angle(
-        ``return_dataset=False`` is useful when ``source_Sv`` is a path and
+    to_disk: bool, default=True
+        If ``True``, ``to_disk`` with split-beam angles added will be returned.
+        ``to_disk=False`` is useful when ``source_Sv`` is a path and
         users only want to write the split-beam angle data to this path.
 
     Returns
     -------
     xr.Dataset or None
-        If ``return_dataset=False``, nothing will be returned.
-        If ``return_dataset=True``, either the input dataset ``source_Sv``
+        If ``to_disk=False``, nothing will be returned.
+        If ``to_disk=True``, either the input dataset ``source_Sv``
         or a lazy-loaded Dataset (from the path ``source_Sv``)
         with split-beam angles added will be returned.
```
I think the True and False conditions here are reverse.

> to_disk=True means that the angle will be saved to disk via mode="a" and the source_Sv dataset lazy-loaded back and returned, and to_disk=False means that the angle will just be added to source_Sv and returned but not saved to disk.

praneethratna replied:
Updated!

leewujung commented on echopype/consolidate/api.py:
```diff
@@ -364,9 +369,19 @@ def add_splitbeam_angle(
+    # add the split-beam angles to the provided Dataset
+    source_Sv["angle_alongship"] = theta
+    source_Sv["angle_athwartship"] = phi
+    if to_disk:
+        if source_Sv_type is not None:
```
I think there should be a check up front, right after `source_Sv_type = get_file_format(source_Sv)`, where you fail the case where `to_disk=True and source_Sv_type is None`, otherwise it is not obvious what's going on if people put `to_disk=True` but the to_netcdf/to_zarr are not invoked due to `source_Sv_type` being None.

leewujung replied:
Actually, looking at `get_file_format`, it does not seem like the output would ever be None. Could you rework the logic up front? What I wanted to say in this and the previous comment is that, if people put in a `source_Sv` that is an in-memory object instead of a path, an error should be raised if they also set `to_disk=True`, because there is no place the data can be saved to.

praneethratna replied:
@leewujung Understood, an error will be raised if `source_Sv` is not a path and `to_disk` is true.

leewujung reviewed (changes requested):
@praneethratna : I put in a couple comments about changes needed in the logic of checking the combination `to_disk` and `souce_Sv` type. Thanks!

leewujung commented on echopype/tests/consolidate/test_consolidate_integration.py:
```diff
@@ -340,7 +340,7 @@ def _tests(ds_test, location_type, nmea_sentence=None):
                 'splitbeam/DY1801_EK60-D20180211-T164025_angles_T5.mat'
             ],
-            "CW", "power", False, True
+            "CW", "power", False, False
```
So this has to be `True` because here we need to test whether both `to_disk=True` and `to_disk=False` are working. This is the only case where `to_disk=True` is tested.

The reason why this test fails right now is because you are doing `source_Sv_type = get_file_format(source_Sv)` after `source_Sv = open_source(source_Sv, "dataset", storage_options)`, so `source_Sv` is no longer a path/str when you do `get_file_formate`.

Good to be careful about changing existing testing parameter combination, since they are usually there for a reason.

Also suggest changing the `write_Sv_to_file` parameter to just `to_disk` to make the code easy to read.

praneethratna replied:
@leewujung Ah, sorry, misunderstood the code point there. Updated now!

leewujung reviewed (changes requested):
@praneethratna : I found a bug in the latest change and commented on it. Should be pretty straightforward to fix.

For the testing case where `to_disk=True`, please also add the below to make sure that it is actually loading data from the temp zarr file (in which case the `.data` is a dask array):
```python
assert isinstance(source_Sv["angle_alongship"].data, dask.array.core.Array)
assert isinstance(source_Sv["angle_athwartship"].data, dask.array.core.Array)
```

Thanks!

leewujung reviewed (approved):
Awesome, thanks for the changes! I'll merge this now!
